framework/src/variables/MooseVariableFV.C

```diff
@@ -463,17 +463,28 @@ MooseVariableFV<OutputType>::getElemValue(const Elem * const elem) const
 
   const dof_id_type index = this->_dof_indices[0];
 
-  ADReal value = (*_solution)(index);
+  // It's not safe to use solutionState(0) because it returns the libMesh System solution member
+  // which is wrong during things like finite difference Jacobian evaluation, e.g. when PETSc
+  // perturbs the solution vector we feed these perturbations into the current_local_solution
+  // while the libMesh solution is frozen in the non-perturbed state
+  const auto & global_soln = (time.state == 0)
+                                 ? *_sys.currentSolution()
+                                 : _sys.solutionState(time.state, time.iteration_type);
```

Comment:
Could we do this under the hood in `solutionState()`?

Reply:
No. The zeroth solution state in `SystemBase::solutionState` is the `internalSolution()` which is not the `currentSolution()` that MOOSE has always used when evaluating residuals and Jacobians. The `currentSolution()` has constraints like hanging nodes and periodic boundaries pre-applied. However, perhaps with the work that @roystgnr is currently doing with adding a libMesh "nonlinear consistent" constraint option, we can remove this distinction in the future. I'd have to think more about it. But as things are now the solutions are different and you can only rely on the `currentSolution()`.